Fix AugmentedMemoizationPolicy max history bug.

kedz opened the pull request:
**Proposed changes**:

The tracker event sequence is truncated at the most recent `ActionExecuted("action_listen")` event before before the `max_history`-th action, preserving the user utterance after the `action_listen` so that it can be correctly included in the tracker state creation. Otherwise, if the number of actions is greater that max_history, `AugmentedMemoizationPolicy` will not be able to recognize them.


**Problem Description**:
Consider the following training story:

```yaml
- story: training example
  steps:
    - intent: customer_service
    - action: action_check_login
    - slot_was_set:
      - login: true
    - action: action_check_queue
    - slot_was_set:
      - routing_queue: "xyz"
    - action: action_check_price_of_doge_coin
    - action: utter_helpful_message1
```

The following training states would be created from the training story above:

```
state1 = {}
state2 = {"prev_action": {"action_name": "action_listen"}, "user": {"intent": "customer_service"}}
state3 = {"prev_action": {"action_name": "action_check_login"}, "user": {"intent": "customer_service"}, "slots": {"login": (1.0, 1.0)}}
state4 = {"prev_action": {"action_name": "action_check_queue"}, "user": {"intent": "customer_service"}, "slots": {"login": (1.0, 1.0), "routing_queue": (0.0, 1.0, 0.0)}}
state5 = {"prev_action": {"action_name": "action_check_price_of_doge_coin"}, "user": {"intent": "customer_service"}, "slots": {"login": (1.0, 1.0), "routing_queue": (0.0, 1.0, 0.0)}}
state6 = {"prev_action": {"action_name": "action_utter_helpful_message1"}, "user": {"intent": "customer_service"}, "slots": {"login": (1.0, 1.0), "routing_queue": (0.0, 1.0, 0.0)}}
```

Assume that `max_history == 3` for the `AugmentedMemoizationPolicy`.
From this training state sequence we would create the following state sequence / action label mappings:

```python
    (state1) - > "action_listen"
    (state1, state2) -> "action_check_login"
    (state1, state2, state3) -> "action_check_queue"
    (state2, state3, state4) -> "action_check_price_of_doge_coin"
    (state3, state4, state5) -> "action_utter_helpful_message1"
    (state4, state5, state6) -> "action_listen"
```

However, at test time, if we had the following event sequence:
```yaml
- action: action_listen
- intent: blah_blah_blah
- action: action_some_bot_action
- action: utter_some_bot_speak
- action: action_listen 
- intent: customer_service
- action: action_check_login
- slot_was_set:
  - login: true
- action: action_check_queue
- slot_was_set:
  - routing_queue: "xyz"
- action: action_check_price_of_doge_coin
```

`AugmentedMemoizationPolicy` first computes where to truncate the events according to the most recent `max_history` events. In this case it would truncate the event sequence to:

```yaml
- action: action_check_login
- slot_was_set:
  - login: true
- action: action_check_queue
- slot_was_set:
  - routing_queue: "xyz"
- action: action_check_price_of_doge_coin
```
This sequence when turned into prediction states, looks like this:

```python
bad_state1 = {"prev_action": {"action_name": "action_check_login"}, "slots": {"login": (1.0, 1.0)}}
bad_state2 = {"prev_action": {"action_name": "action_check_queue"}, "slots": {"login": (1.0, 1.0), "routing_queue": (0.0, 1.0, 0.0)}}
bad_state3 = {"prev_action": {"action_name": "action_check_price_of_doge_coin"}, "slots": {"login": (1.0, 1.0), "routing_queue": (0.0, 1.0, 0.0)}}
``` 
Crucially, the user intent information is lost so looking up `(bad_state1, bad_state2, bad_state3)` yields a miss and policy prediction is handed off to TED.

Instead, this PR changes the truncation behavior, to truncate the tracker events on the most recent `action_listen `before where `max_history` would originally imply truncating. For example:

```yaml
- action: action_listen
- intent: blah_blah_blah
- action: action_some_bot_action
- action: utter_some_bot_speak
- action: action_listen       # <- where this PR now truncates
- intent: customer_service
- action: action_check_login         # <- where to truncate according to `max_history`
- slot_was_set:
  - login: true
- action: action_check_queue
- slot_was_set:
  - routing_queue: "xyz"
- action: action_check_price_of_doge_coin
```

Even though this new event sequence contains more than `max_history` events, when handed off to the `MaxHistoryTrackerFeaturizer`, it will first create the states (with the correct user info) and then truncate to `max_history` states, which `AugmentedMemoizationPolicy` will use to lookup.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] ~updated the documentation~
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

kedz commented:
/modeltest

```yml
 
include:
  - dataset: ["all"]
    config: ["all-core"]
```

samsucik commented:
@kedz I'm re-running the regression tests after making some fixes to the training data repo on the `rasa-v2.x` branch ([this commit](https://github.com/RasaHQ/training-data/commit/d20717b4d0a653ec33d0a189ef81b2fb291b8443)) to correctly use 2.x domain files. This should fix the failing regression tests and produce meaningful numbers 🙂

samsucik commented on changelog/10606.bugfix.md:
```diff
@@ -0,0 +1 @@
+Fix `max_history` truncation in `AugmentedMemoizationPolicy` to preserve the most recent `UserUttered` event.
```
I'd also explain what the effect of this bug was, e.g.
```suggestion
Fix `max_history` truncation in `AugmentedMemoizationPolicy` to preserve the most recent `UserUttered` event.
Previously, `AugmentedMemoizationPolicy` failed to predict next action after long sequences of actions (longer than `max_history`) because the policy did not have access to the most recent user message.
```

samsucik reviewed (approved):
The fix seems straightforward 🚀 I suggested a few non-substantial improvements. Also, let's wait until the regression tests have finished.

Out of curiosity: Does this bug affect 3.x as well? I'm thinking not but I also don't know much about the relevant changes in 3.x (maybe @ka-bu will know). If this affected 3.x too, then an interesting question is how to handle the fix -- usually, we'd fix things in the newer version and back-port, but maybe this doesn't work if we want to release fixes for customers in an older version very quickly? 🤔

kedz commented:
/modeltest

dataset_branch: "rasa-v2.x" 
include:
  - dataset: ["Sara"]
    config: ["Rules + AugMemo"]

kedz commented:
/modeltest

```yaml
dataset_branch: "rasa-v2.x" 
include:
  - dataset: ["Sara"]
    config: ["Rules + AugMemo"]
```

kedz commented on tests/core/policies/test_memoization.py:
```diff
@@ -157,3 +157,89 @@ def test_augmented_prediction(self, max_history):
+                ActionExecuted(UTTER_GREET_ACTION),
+                UserUttered(intent={"name": GREET_INTENT_NAME}),
+                ActionExecuted(UTTER_GREET_ACTION),
+                SlotSet("slot_2", True),
```
Just that augmemo would disregard the context. I have simplified this in the new version.

kedz commented on tests/core/policies/test_memoization.py:
```diff
@@ -157,3 +157,89 @@ def test_augmented_prediction(self, max_history):
         )
+
+    @pytest.mark.parametrize("max_history", [1, 2, 3, 4, None])
+    def test_augmented_prediction_long_story(self, max_history):
```
Added a docs string and renamed. Hopefully it is more clear now.

samsucik replied:
Awesome!

kedz commented on tests/core/policies/test_memoization.py:
```diff
@@ -157,3 +157,89 @@ def test_augmented_prediction(self, max_history):
         )
+
+    @pytest.mark.parametrize("max_history", [1, 2, 3, 4, None])
+    def test_augmented_prediction_long_story(self, max_history):
```
Thanks for these suggestions! i added additional tests and also because of them ended up finding another bug in augmemo, that is similar to but different than the one I was trying to fix:

Originally we have a test for the following situation, a training tracker that starts with an action:

```python
        training_story = TrackerWithCachedStates.from_events(
            "training story",
            [
                ActionExecuted(UTTER_GREET_ACTION),
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_3", True),
                ActionExecuted(UTTER_BYE_ACTION),
            ],
            domain=domain,
            slots=domain.slots,
        )
```

this will match the following test story because even though there are unrelated events at the start AugmentedMemoizationPolicy will keep popping initial actions off the tracker until it either exhausts the tracker or it finds a match:

```python
        test_story = TrackerWithCachedStates.from_events(
            "test story",
            [
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_1", False),
                ActionExecuted(UTTER_GREET_ACTION),
                ActionExecuted(UTTER_GREET_ACTION),
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_2", True),
                ActionExecuted(UTTER_GREET_ACTION),
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_3", True),
                # ActionExecuted(UTTER_BYE_ACTION),
            ],
            domain=domain,
            slots=domain.slots,
        )
```        
        
however if the training story starts with a UserUttered event it will not find a match, e.g.:

```python
        training_story = TrackerWithCachedStates.from_events(
            "training story",
            [
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_3", True),
                ActionExecuted(UTTER_BYE_ACTION),
            ],
            domain=domain,
            slots=domain.slots,
        )
```
because the test story will get truncated to after the UserUttered event, which change the underlying states. 

Not that this is not about tracker truncation which happens [here](https://github.com/RasaHQ/rasa/blob/a9d885f20add8f2bdcc5b65b6863555dd68c9fc5/rasa/core/policies/memoization.py#L401) and that I had fixed. This new issue happens [here](https://github.com/RasaHQ/rasa/blob/a9d885f20add8f2bdcc5b65b6863555dd68c9fc5/rasa/core/policies/memoization.py#L290) in the `_back_to_the_future` method. This method originally tried to truncate the tracker at the next action, effectively removing user utterance information from the state.  I was able to fix it by making the `_back_to_the_future` method truncate to the next action **or user uttered event.** This will add extra checks but since its only user uttereances and only one of those can happen per arbitrarily many bot actions I think it will be negligible.

samsucik replied:
@kedz I admit that I haven't looked at the code (following your links) but I've verified that, even if a training story starts with an intent, it gets translated into a tracker (and subsequently into states) where there's `action_listen` before the intent. So, I think the scenario you described can't happen today. In any case, AugMemo has lots of "magical" code that is loaded with assumptions and we could at least make sure this one (stories starting with actions) is made clear in there (or is even checked for?). Wdyt?

kedz commented:
@samsucik 

> Out of curiosity: Does this bug affect 3.x as well? I'm thinking not but I also don't know much about the relevant changes in 3.x (maybe @ka-bu will know). If this affected 3.x too, then an interesting question is how to handle the fix -- usually, we'd fix things in the newer version and back-port, but maybe this doesn't work if we want to release fixes for customers in an older version very quickly? 

It does affect 3.x as well. Working on a PR for that now. (I also found a another bug in the 3.0 tests unrelated to the ones here 🤦🏻 )

kedz commented:
/modeltest

```yaml
dataset_branch: "rasa-v2.x" 
include:
  - dataset: ["retail-demo", "Sara"]
    config: ["all-core"]
```

kedz commented:
@samsucik @JEM-Mosig regression tests might be working here (but still kinda flaky) but the full sweet ran without issue on the 3.0 PR [here](https://github.com/RasaHQ/rasa/pull/10618).

samsucik commented on tests/core/policies/test_memoization.py:
```diff
@@ -157,3 +227,306 @@ def test_augmented_prediction(self, max_history):
+        )
+
+    @pytest.mark.parametrize("max_history", [1, 2, 3, 4, None])
+    def test_aug_pred_connects_different_memoizations(self, max_history):
```
Great test case! Some clean-up could be done by removing the unused slots (`slot_2`, `slot_3`) and adding `action_listen` before user utterances.

samsucik reviewed (changes requested):
The bulk of the changes are good to go; I left a few comments and questions 🙂

samsucik commented on .github/workflows/ci-model-regression.yml:
```diff
@@ -643,7 +643,7 @@ jobs:
           SCHEDULE_ID=$(curl -X GET -s -H 'Authorization: token ${{ secrets.GITHUB_TOKEN }}' -H "Accept: application/vnd.github.v3+json" \
             "https://api.github.com/repos/${{ github.repository }}/actions/workflows" \
-            | jq -r  '.workflows[] | select(.name == "${{ github.workflow }}") | select(.path | test("schedule")) | .id')
+            | jq -r  '.workflows[] | select(.name == "CI - Model Regression on schedule") | select(.path | test("schedule")) | .id')
```
And is this something that should differ between 2.8 and 3.0? Since I don't see this change on the 3.0 PR...

kedz replied:
This is what markus told me was the only difference between 2.8 and 3.0 in terms of the workflow [here](https://rasa-hq.slack.com/archives/C02HT93BDTM/p1640881364006900). I'm not sure if this actually fixes anything (but the reg tests appear to be working now). This is how the workflow is on 3.0.x now.

samsucik replied:
Alright, I see. A tiny change but I think definitely worth mentioning somewhere -- not in the changelog as it's an internal thing (reg. tests) but at least in the merging commit's message or this PR's name? 🤔  Other than that, we can resolve this thread.

JEM-Mosig commented:
@kedz Great analysis in the issue! But I'm a bit worried that we make the meaning of `max_history` even more fuzzy with this. For all other policies `max_history` refers to the number of actions (including `action_listen`) to take into account. This may already be confusing, but with this change it would be different for just AugMemo. Can one not just increase the `max_history` parameter to 20 or something to resolve the issue?

ka-bu commented:
> @kedz Great analysis in the issue! But I'm a bit worried that we make the meaning of `max_history` even more fuzzy with this. For all other policies `max_history` refers to the number of actions (including `action_listen`) to take into account. 

In the `MaxHistoryTrackerFeaturizer` the `max_history` refers to the states, which should correspond to actions (!= action_listen), no? So the meaning here seems consistent to me (given that the `_prediction_states` uses the max history featurizer under the hood which will get rid of any extra states obtained by including `ActionExecuted` events between the first `ActionExecuted` inside the `max_history` window and the `ActionListen` before that). Or are you thinking of something else here?

> confusing

What I found really confusing is that the context is also pruned within the max history window. I would've assumed that AugMemo just cuts off history prior to the `max_history` boundary. But the truncation happens successively and the fix undoes that "just a little bit" each time to just include the prior utterance. ... 

> Can one not just increase the `max_history` parameter to 20 or something to resolve the issue?

Yep, seems to me like it would solve the example above. But I think one problem that we can't get rid of because it happens by design is that: If we want our lookup table to contain combinations of "older" slots and recent actions, then we can't achieve that by increasing the history because aug memo will successively cutoff every prefix *inside* that window. 
... Is that also an issue (and wouldn't the approach with "cutoff everything before `max_history` and then behave like `MaxHistoryTrackerFeaturizer` be an option then)? 🤔  @kedz ?

samsucik commented:
Great points, Johannes and Kathrin! I'll let Chris react but I just wanted to highlight that this issue has now definitely out-grown an on-point task and we should make this more transparent -- e.g. by properly determining the priority of the issue and/or assigning it to someone who is on the reactive team (unless Chris wants to keep it, of course!).

kedz commented on tests/core/policies/test_memoization.py:
```diff
@@ -89,7 +90,76 @@ def create_policy(
+        training_story = TrackerWithCachedStates.from_events(
+            "training story",
+            [
+                UserUttered(intent={"name": GREET_INTENT_NAME}),
```
I rewrote the trackers to be more realistic (action_listens precede user utterances, and always start and end training sequences.

kedz commented on tests/core/policies/test_memoization.py:
```diff
@@ -89,7 +90,76 @@ def create_policy(
+        )
+
+    @pytest.mark.parametrize("max_history", [1, 2, 3, 4, None])
+    def test_augmented_prediction_when_starts_with_action(self, max_history):
```
I consolidated this and with_intent into one test with a realistic tracker.

kedz commented:
@samsucik I rewrote the tests with more realistic trackers (action_listen precedes user utterances and is the final action), and renamed some private methods at your suggestion. If we are expecting trackers to be realistic we don't need the changes I made to `_strip_leading_events_until_action_executed` (formerly `_back_to_the_future` 😱 ) (the change being to try truncating that tracker at both actions and user uttered events), so I reverted those changes.

@JEM-Mosig 
> But I'm a bit worried that we make the meaning of max_history even more fuzzy with this. For all other policies max_history refers to the number of actions (including action_listen) to take into account. 

I would agree with @ka-bu that `max_history` is referring to the number of states considered which this PR doesn't change. It only extends the tracker event history truncation marginally so that the states are created correctly before `max_history` state sequence truncation is performed. 

The tracker event history truncation was only introduced because of [this issue](https://github.com/RasaHQ/rasa-x/issues/5546) to handle the case where in production a tracker might grow arbitrarily long and it was slowing down inference time significantly; `max_history` actions would be the minimum we could truncate to and preserve the correct behavior (modulo this bug) but where we truncate the tracker event sequence, assuming it doesn't hurt inference time noticeably, is somewhat arbitrary. 


I suspect that increasing the `max_history` of AugMemo would be a solution but then it creates a discrepancy between how states are created at training time and test time, which could be confusing to us and others. I think not fixing this bug and just saying to extend the window as a solution would warrant a docs update to say that your `max_history` window should be at least as long as the longest number of actions between user responses + 1 (a rule I suspect many will forget even if they read the docs). But if everyone else feels like going this route, I can create the docs modification and revert these changes.

Let me know on how I should proceed.

@ka-bu  Regarding this:
> But I think one problem that we can't get rid of because it happens by design is that: If we want our lookup table to contain combinations of "older" slots and recent actions, then we can't achieve that by increasing the history because aug memo will successively cutoff every prefix inside that window.

I think that AugMemo is intended to be use explicitly to ignore older slots. But maybe I am not understanding exactly what you mean here.

ka-bu commented:
> @ka-bu Regarding this:
> 
> > But I think one problem that we can't get rid of because it happens by design is that: If we want our lookup table to contain combinations of "older" slots and recent actions, then we can't achieve that by increasing the history because aug memo will successively cutoff every prefix inside that window.
> 
> I think that AugMemo is intended to be use explicitly to ignore older slots. But maybe I am not understanding exactly what you mean here.

Not sure this is really relevant, but I would've expected the  "ignore older slots" to only mean: ignore slot sets that happened *more* than `max_history` states before. But it seems to me that we also remove slot sets within that `max_history` window, because we  `states = self._prediction_states(truncated_tracker, domain,)` with the `truncated_tracker` from which we remove the leading events. So I think we won't have a lookup for "(slot set that happened x steps ago inside `max_history` window but more than `n` states before current state) + (last `m` states with `m <<n`" via AugMemo, no?

kedz commented:
@ka-bu ah ok I think I understand now. Yeah this doesn't change the behavior within the `max_history` window of states. Slots will be progressively be discarded, regardless of whether there is a matching prefix in the training that would keep them. I think that is out of scope for this issue but point me to that notion page you mentioned. Sounds interesting.

JEM-Mosig commented on rasa/core/policies/memoization.py:
```diff
@@ -342,10 +351,12 @@ def _recall_using_delorean(
+            truncated_tracker
+        )
+        while truncated_tracker is not None:
+            states = self._prediction_states(truncated_tracker, domain,)
```
```suggestion
            states = self._prediction_states(truncated_tracker, domain)
```

JEM-Mosig commented on rasa/core/policies/memoization.py:
```diff
@@ -286,28 +287,36 @@ class AugmentedMemoizationPolicy(MemoizationPolicy):
         applied_events = tracker.applied_events()
 
-        # we need to find second executed action
+        # We need to find the second `ActionExecuted` or `UserUttered` event.
```
This is not true. The code below finds the second ActionExecuted, no?

kedz replied:
Ah forget to change that back. Reverted to original.

JEM-Mosig commented on rasa/core/policies/memoization.py:
```diff
@@ -159,7 +160,7 @@ def _create_feature_key(self, states: List[State]) -> Text:
         # represented as dictionaries have the same json strings
         # quotes are removed for aesthetic reasons
         feature_str = json.dumps(states, sort_keys=True).replace('"', "")
-        if self.ENABLE_FEATURE_STRING_COMPRESSION:
+        if False:  # self.ENABLE_FEATURE_STRING_COMPRESSION:
```
Why remove this in this PR?

kedz replied:
🤦🏻 ah thanks for catching that. Was looking at how the states got cached when writing tests and forgot to unset. Reverted now.

JEM-Mosig reviewed (approved):
Looks good to me! Just some minor things.

samsucik commented on tests/core/policies/test_memoization.py:
```diff
@@ -422,10 +351,9 @@ def test_aug_pred_sensitive_to_intent_across_max_history_actions(self, max_histo
         )
```
The asserted `action_listen` in the above line is something new to me -- @kedz do you know where it's coming from? I'm just curious because here it looks as if listening was the default action but it's the first time I've encountered this design decision...

kedz replied:
I think I should change the test because this actually not the default behavior but falling out of how the test checks for the action name.

in this case `prediction2.probabilities == [0.0, 0.0, ..., 0.0]` so the argmax of it is index 0 which is usually `action_listen`, but I don't know if we guarantee it to be so. I'll mod the test to assert that all probabilities are zero.

samsucik replied:
Aaaah, I see! Alrighty, the changes you've made make a complete sense to me 🙂

samsucik reviewed (approved):
Let's go 🚀 

> I reverted those changes.

Good catch!

>  `max_history` is referring to the number of states considered which this PR doesn't change. It only extends the tracker event history truncation marginally so that the states are created correctly

I also agree with this.

> I think that AugMemo is intended to be use explicitly to ignore older slots.

That's my understanding as well, from talking to Vova long ago.

As for the issue that Kathrin pointed out (slots getting truncated too): It's definitely not an obvious choice and could be challenged, but I agree that it's outside the scope of this issue.

samsucik commented:
@kedz could you run the relevant reg. tests one last time, if/once you're done adding commits?
